test/functional/test_framework/util.py

```diff
@@ -387,32 +390,125 @@ def ensure_for(*, duration, f, check_interval=0.2):
+                        results = list(executor.map(lambda p: p(), predicate))
+                        all_true = all(results)
+                else:
+                    if lock:
+                        with lock:
+                            result = predicate()
+                    else:
+                        result = predicate()
+                    results = [result]
+                    all_true = result
+            elif isinstance(predicate, str):
+                try:
+                    subprocess.check_call(predicate, shell=True)
+                    result = True
+                except subprocess.CalledProcessError:
+                    result = False
+                results = [result]
+                all_true = result
+            else:
+                raise ValueError("Predicate must be callable, list of callables, or str (shell command).")
+
+            loop_history.append((time.time(), results))
+            if all_true:
+                local_history.extend(loop_history)
                 return
-        time.sleep(check_interval)
 
-    # Print the cause of the timeout
-    predicate_source = "''''\n" + inspect.getsource(predicate) + "'''"
-    logger.error("wait_until() failed. Predicate: {}".format(predicate_source))
-    raise AssertionError("Predicate {} not true after {} seconds".format(predicate_source, timeout))
+            time.sleep(check_interval)
+
+        local_history.extend(loop_history)
+        attempt += 1
+        if attempt <= retries:
+            check_interval *= 1.2
+            current_timeout *= 1.1
+
+    predicate_source = inspect.getsource(predicate) if callable(predicate) else str(predicate)
+    call_stack = '\n'.join([f"{frame.filename}:{frame.lineno} {frame.function}" for frame in inspect.stack()])
+    memory_usage = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024
+    cpu_times = os.times()
+
+    error_msg = (
+        f"wait_until() failed after {timeout} seconds (adjusted: {current_timeout}, retries: {retries}).\n"
+        f"Predicate: ''''\n{predicate_source}\n''''\n"
+        f"Call stack:\n{call_stack}\n"
+        f"Memory usage: {memory_usage:.2f} MB\n"
+        f"CPU times: {cpu_times}\n"
+    )
+    logger.error(error_msg)
+
+    # Analyze history
+    if local_history:
+        times = [t for t, _ in local_history]
+        if len(times) > 1:
+            deltas = [times[i + 1] - times[i] for i in range(len(times) - 1)]
+            avg_delta = sum(deltas) / len(deltas)
+            suggested_interval = avg_delta * 1.5
+            logger.info(f"Suggested check_interval adjustment: {suggested_interval:.3f} (based on average poll delta {avg_delta:.3f})")
+
+        num_fails = sum(
+            1 for _, r in local_history
+            if (not all(r) if isinstance(r, list) else not r)
+        )
+        logger.info(f"AI-assisted insight: Predicate failed {num_fails}/{len(local_history)} times. Possible flaky timing issue.")
```

Comment:
Human insight: please don't open PRs like this. None of this seems actually useful and if it were you should be able to point to where these changes to the framework do actually improve the test suite.